codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/CodegenEngine.java

```diff
@@ -374,4 +388,127 @@ public enum TargetLanguage {
+                            "aws.polymorph#dafnyUtf8Bytes",
+                            "aws.polymorph#javadoc",
+                            "aws.polymorph#positional",
+                            "aws.polymorph#reference"
+                    )).toList();
+
+    private static final List<String> SUPPORTED_TRAITS_AWS_SDK_STYLE =
+            Stream.concat(
+                    COMMON_SUPPORTED_TRAITS.stream(),
+                    Stream.of(
+                            // Most of these are protocol details handled by the wrapped SDKs
+                            // and not relevant for SDK consumers.
+                            "aws.api#clientEndpointDiscovery",
+                            "aws.api#clientDiscoveredEndpoint",
+                            "aws.api#service",
+                            "aws.auth#sigv4",
+                            "aws.protocols#awsJson1_0",
+                            "aws.protocols#awsJson1_1",
+                            "aws.protocols#awsQuery",
+                            "aws.protocols#awsQueryError",
+                            "smithy.api#deprecated",
+                            "smithy.api#paginated",
+                            "smithy.api#suppress",
+                            "smithy.api#httpError",
+                            "smithy.api#title",
+                            "smithy.api#xmlFlattened",
+                            "smithy.api#xmlName",
+                            "smithy.api#xmlNamespace",
+                            "smithy.rules#endpointTests",
+                            "smithy.rules#endpointRuleSet",
+                            // We don't really support this yet, since it implies extra API
+                            // methods we don't generate, but at least we don't generate incorrect code.
+                            "smithy.waiters#waitable"
+                    )).toList();
+
+    private void checkForUnsupportedFeatures(Collection<String> supportedShapes, Collection<String> supportedTraits) {
+        String selectorExpr =
+                "[id=" + serviceShape.getId() + "] \n" +
+                        ":is(*, ~> *) \n" +
+                        ":not(\n" +
+                            ":is(" + String.join(", ", supportedShapes) + ") " +
+                            "$supportedTraits(:root([id = " + String.join(", ", supportedTraits) + "])) " +
+                            "[@: @{trait|(keys)} {<} @{var|supportedTraits|id}]" +
+                        ")";
+
+        Set<Shape> notSupported = Selector.parse(selectorExpr).select(model);
+
+        if (!notSupported.isEmpty()) {
+            String message = "The following shapes in the service's closure are not supported: \n" +
+                    notSupported.stream()
+                                .map(shape -> unsupportedShapeLine(shape, supportedShapes, supportedTraits))
+                                .collect(Collectors.joining("\n"));
+            throw new IllegalArgumentException(message);
```

Comment:
Note to self: it will be much nicer to make these `DANGER`-level Validator events instead. That means we'll get an event for every unsupported thing, and the ability to suppress them if a user checks that the trait is not relevant or the lack of support is harmless.

There is even precedent for individual code generators providing validators to check for their own limitations:
* Rust: https://github.com/smithy-lang/smithy-rs/blob/14ccc50f41ea8b59199fecf239131826770fa0d1/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/PatternTraitEscapedSpecialCharsValidator.kt#L16
* Typescript: https://github.com/smithy-lang/smithy-typescript/blob/3c9c07ba1784e1b0a04bb44a09890335378ea21b/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/validation/LongValidator.java#L38

Reply:
Note that some things we don't support should be hard `ERROR` events, which are for fatal errors that prevent processing a model and can't be suppressed. We don't support Smithy 2.0 `enum` shapes in .NET for example, and that can't be worked around.